Add release command to setup.py

zachyee opened the pull request:
In order to automate releasing to github and pypi

Currently uses basic authentication (username/password) to release to github. An alternative would be to use OAuth tokens: https://developer.github.com/v3/oauth/

It also currently can only release to pypitest, but adding pypi to `.pypirc` will allow it to release to pypi.
It uploads a source distribution (default for unix is .tar.gz).

ebd2 commented on release.py:
```diff
@@ -0,0 +1,440 @@
+
+    @staticmethod
+    def get_latest_release():
+        latest_release = subprocess.check_output(['curl', '--silent', REPOSITORY_API + '/releases/latest'])
```
Why not urllib2 or something like that?

ebd2 commented on release.py:
```diff
@@ -0,0 +1,440 @@
+
+    @staticmethod
+    def check_branch_up_to_date(local_branch):
+        status = subprocess.check_output(['git', 'status'])
```
Same question but for the git module

ebd2 replied:
See https://github.com/kokosing/docker-release/blob/master/docker_release/main.py for an example

ebd2 commented on setup.py:
```diff
@@ -109,14 +110,15 @@ def get_custom_rc_file(self):
         'Development Status :: 2 - Pre-Alpha',
         'Intended Audience :: Developers',
-        'License :: OSI Approved :: APLv2 License',
+        'License :: OSI Approved :: Apache Software License',
```
Dumb question from the ignorant: What is APLv2 and Apache Software License? Neither of them appear listed here: https://www.apache.org/licenses/

@mattsfuller: Can you confirm what this should be? You spend more time talking to the lawyers than we do.

zachyee replied:
Would be good to confirm. Also here's a list of valid classifiers: https://pypi.python.org/pypi?%3Aaction=list_classifiers

mattsfuller replied:
It should be Apache License Version 2

ebd2 commented:
Looks basically OK to me, modulo the question about about going through subprocess rather than python modules and the license question.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,440 @@
+except ImportError:
+    from distutils.core import Command
+
+REPOSITORY_API = 'https://api.github.com/repos/zachyee/presto-admin'
```
Testing URL? Change before pushing to master?

ebd2 commented on release.py:
```diff
@@ -0,0 +1,490 @@
+DIST_DIRECTORY = os.path.join(CURRENT_DIRECTORY, 'dist')
+
+
+class release(Command):
```
Most of the methods in here are static. Maybe it makes sense to break out github and pypi files that handle the inner workings of pushing to github and pypi and invoke that functionality here? That might also have the benefit of helping to make clear what the public interfaces are and what functions are internal implementation details.

zachyee commented:
Refactored into separate classes

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+            # The field value was omitted for the version
+            return 0
+
+    def _set_semantic_version_fields(self):
```
Nit: this should probably be inline in the constructor. It isn't used elsewhere, and it's weird that it operates by side-effect like this.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+CURRENT_DIRECTORY = os.path.dirname(os.path.realpath(__file__))
+
+
+class SemanticVersionHelper(object):
```
The object represents a specific version. Proposed name: SemanticVersion

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+        self.minor_version = self._get_version_field_value(version_fields, 1)
+        self.patch_version = self._get_version_field_value(version_fields, 2)
+
+    def __lt__(self, other):
```
Please imlement __eq__ and use @total_ordering on the class.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+
+
+class ReleaseFetcher:
+    def __init__(self, directory, github_api_directory):
```
suggest: github_api_path; it isn't really a directory in the sense that it's a container for files.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+        for line in contents.splitlines():
+            if 'version' in line and '=' in line:
+                if tag_name not in line:
+                    exit('Version has not been updated to %s in %s' % (tag_name, file_path))
```
Include a line number to aid problem solving. The check for a version is a little coarse, which is OK for internal tools, but we should make it easy to know what the problem is if this code finds one.

ebd2 replied:
Actually, it would probably be preferable to extract a version constant in setup.py, and then `from prestoadmin import version` and `from setup import version`, which lets us rely on the python parser.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+        self.is_draft = 'false'
+        self.is_prerelease = 'false'
+
+    def _user_set_username(self):
```
Here and below, suggest name `prompt_username`

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+        self.release_name = 'Release ' + self.tag_name
+
+    @staticmethod
+    def _make_multiline_string_json_friendly(multiline_string):
```
suggest `escape_newlines` for the name.

ebd2 commented on release.py:
```diff
@@ -0,0 +1,488 @@
+        else:
+            return False
+
+    def _send_post_request(self, url, data, authorization_string, content_type, content_length):
```
The GET/POST stuff probably belongs in a new module under `prestoadmin/util`. It's more general than just talking to github.

zachyee commented:
Addressed comments.

Created a util directory with `util/semantic_version.py` and `util/http.py`.

Created `prestoadmin/_version.py` which holds the current version. Unfortunately, setup.py should not import directly from `prestoadmin` because it may (and does) contain third party dependencies. Instead it reads the file, so it doesn't have to be parsed. This is a workaround other modules use, even though it's not pretty. Comments were added to note this.

ebd2 commented:
Looks good to me except for the license.
